worker/src/lib.rs

```diff
@@ -0,0 +1,255 @@
+use std::{
+    io,
+    process::{ExitStatus, Stdio},
+    sync::Arc,
+};
+
+use bytes::Bytes;
+use futures::{future::FusedFuture, FutureExt, Stream};
+use log::error;
+use tokio::{
+    io::BufReader,
+    process::{Child, Command},
+    sync::{oneshot, Mutex, Notify, RwLock},
+};
+
+mod logs;
+mod ops;
+
+async fn process_task(
+    mut child: Child,
+    inner: Arc<ProcessInner>,
+    stop_receiver: oneshot::Receiver<()>,
+) {
+    // A child process might close both stdout and stderr,
+    // but remain alive. In that case, we must still try to wait
+    // for the stop message.
+    // Fuse the future to be able to .await twice safely (when !is_terminated()).
+    let mut stop_receiver = stop_receiver.fuse();
+
+    let stdout = BufReader::new(child.stdout.take().expect("should always be available"));
+    let stderr = BufReader::new(child.stderr.take().expect("should always be available"));
+
+    // Phase 1: copy logs from stdout/stderr, on stop message: signal the child.
+    tokio::select! {
+        copied = futures::future::join(
+            logs::copy(stdout, inner.clone()),
+            logs::copy(stderr, inner.clone())
+        ) => {
+            if let Err(e) = copied.0 {
+                error!("{:?}", e);
+            }
+            if let Err(e) = copied.1 {
+                error!("{:?}", e);
+            }
+        },
+        _ = &mut stop_receiver => {
+            if let Err(e) = ops::stop_child(&mut child).await {
+                error!("{:?}", e);
+            }
+        },
+    };
+
+    // Phase 2: stdout/stderr have been closed,
+    // wait for a stop message to arrive (if not arrived yet),
+    // or on the child to finish otherwise
+    loop {
+        tokio::select! {
+            _ = &mut stop_receiver, if !stop_receiver.is_terminated() => {
+                if let Err(e) = ops::stop_child(&mut child).await {
+                    error!("{:?}", e);
+                }
+            },
+            res = child.wait() => {
+                match res {
+                    Ok(s) => {
+                        inner.finish(s).await;
+                        inner.progress.notify_waiters();
+                        return;
+                    }
+                    Err(_) => unimplemented!("when does this happen?"), // TODO
+                }
+            }
+        }
+    }
+}
+
+pub(crate) struct ProcessInner {
+    exit_status: RwLock<Option<ExitStatus>>,
+    logs: RwLock<Vec<Bytes>>,
```

Comment:
If we don't do a lot of chaining and appending of bytes as well as splits and shares, using the library is rather pointless and you may be better served by `Vec<u8>`.

Reply:
@xacrimon My thinking was, this would be a bit more efficient considering the server component. Multiple clients might be streaming the process's logs, with `Bytes` cloning the underlying bytes would be avoided.

Reply:
Ah, you're correct. I believe bytes has some other overheads though but not sure on that. Thanks for the explanation, feel free to mark as resolved.